File: sdk/csharp/Rin.Client/HostDurability.cs
namespace Rin.Client;

public enum HostDurabilityProfile
{
    Advisory = 0,
    IdempotentAction = 1,
    TransactionalAction = 2,
}

public sealed record HostDurability(
    int Version,
    HostDurabilityProfile Profile,
    bool StableIdentity,
    bool DurableBeforeNetwork,
    bool DurableOutbox,
    bool IdempotentApply,
    bool AtomicApplyAndOutbox)
{
    public static HostDurability Advisory(bool stableIdentity = false) =>
        new(1, HostDurabilityProfile.Advisory, stableIdentity, false, false, false, false);

    public static HostDurability IdempotentAction() =>
        new(1, HostDurabilityProfile.IdempotentAction, true, true, true, true, false);

    public static HostDurability TransactionalAction(bool idempotentApply = false) =>
        new(1, HostDurabilityProfile.TransactionalAction, true, true, true, idempotentApply, true);

    public HostDurability Validate()
    {
        if (Version != 1 || !Enum.IsDefined(typeof(HostDurabilityProfile), Profile))
        {
            throw new RinConfigurationException(
                "invalid_host_durability",
                "Host durability has an unsupported version or profile");
        }
        if (Profile == HostDurabilityProfile.IdempotentAction &&
            !(StableIdentity && DurableBeforeNetwork && DurableOutbox && IdempotentApply))
        {
            throw new RinConfigurationException(
                "invalid_host_durability",
                "IdempotentAction requires stable durable state, Outbox, and idempotent apply");
        }
        if (Profile == HostDurabilityProfile.TransactionalAction &&
            !(StableIdentity && DurableBeforeNetwork && DurableOutbox && AtomicApplyAndOutbox))
        {
            throw new RinConfigurationException(
                "invalid_host_durability",
                "TransactionalAction requires stable durable state, Outbox, and atomic settlement");
        }
        return this;
    }

    public void Require(HostDurabilityProfile requiredDurability)
    {
        Validate();
        if (!Enum.IsDefined(typeof(HostDurabilityProfile), requiredDurability))
        {
            throw new RinConfigurationException(
                "invalid_host_durability_profile",
                "Required host durability profile is unknown");
        }
        if ((int)Profile < (int)requiredDurability)
        {
            throw new RinConfigurationException(
                "host_durability_insufficient",
                $"Action requires {Label(requiredDurability)}, but host provides {Label(Profile)}");
        }
    }

    public static string Label(HostDurabilityProfile profile) => profile switch
    {
        HostDurabilityProfile.Advisory => "advisory",
        HostDurabilityProfile.IdempotentAction => "idempotent-action",
        HostDurabilityProfile.TransactionalAction => "transactional-action",
        _ => throw new RinConfigurationException(
            "invalid_host_durability_profile",
            "Host durability profile is unknown"),
    };
}

File: sdk/java/src/main/java/io/github/sunrioa/rin/HostDurability.java
package io.github.sunrioa.rin;

public record HostDurability(
        int version,
        HostDurabilityProfile profile,
        boolean stableIdentity,
        boolean durableBeforeNetwork,
        boolean durableOutbox,
        boolean idempotentApply,
        boolean atomicApplyAndOutbox) {

    public HostDurability {
        if (version != 1 || profile == null) {
            throw new RinConfigurationException(
                    "invalid_host_durability",
                    "Host durability has an unsupported version");
        }
        if (profile == HostDurabilityProfile.IDEMPOTENT_ACTION &&
                !(stableIdentity && durableBeforeNetwork && durableOutbox && idempotentApply)) {
            throw new RinConfigurationException(
                    "invalid_host_durability",
                    "idempotent-action requires stable durable state, Outbox, and idempotent apply");
        }
        if (profile == HostDurabilityProfile.TRANSACTIONAL_ACTION &&
                !(stableIdentity && durableBeforeNetwork && durableOutbox && atomicApplyAndOutbox)) {
            throw new RinConfigurationException(
                    "invalid_host_durability",
                    "transactional-action requires stable durable state, Outbox, and atomic settlement");
        }
    }

    public static HostDurability advisory() {
        return advisory(false);
    }

    public static HostDurability advisory(boolean stableIdentity) {
        return new HostDurability(
                1,
                HostDurabilityProfile.ADVISORY,
                stableIdentity,
                false,
                false,
                false,
                false);
    }

    public static HostDurability idempotentAction() {
        return new HostDurability(
                1,
                HostDurabilityProfile.IDEMPOTENT_ACTION,
                true,
                true,
                true,
                true,
                false);
    }

    public static HostDurability transactionalAction(boolean idempotentApply) {
        return new HostDurability(
                1,
                HostDurabilityProfile.TRANSACTIONAL_ACTION,
                true,
                true,
                true,
                idempotentApply,
                true);
    }

    public void require(HostDurabilityProfile requiredDurability) {
        if (requiredDurability == null) {
            throw new RinConfigurationException(
                    "invalid_host_durability_profile",
                    "Required host durability profile is unknown");
        }
        if (profile.rank() < requiredDurability.rank()) {
            throw new RinConfigurationException(
                    "host_durability_insufficient",
                    "Action requires " + requiredDurability.label() +
                            ", but host provides " + profile.label());
        }
    }
}
